pinot-core/src/main/java/org/apache/pinot/core/operator/filter/SortedIndexBasedFilterOperator.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.filter;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import org.apache.pinot.common.utils.Pairs.IntPair;
+import org.apache.pinot.core.common.DataSource;
+import org.apache.pinot.core.io.reader.impl.v1.SortedIndexReader;
+import org.apache.pinot.core.operator.blocks.FilterBlock;
+import org.apache.pinot.core.operator.docidsets.SortedDocIdSet;
+import org.apache.pinot.core.operator.filter.predicate.PredicateEvaluator;
+import org.apache.pinot.core.operator.filter.predicate.RangePredicateEvaluatorFactory.OfflineDictionaryBasedRangePredicateEvaluator;
+
+
+@SuppressWarnings("rawtypes")
+public class SortedIndexBasedFilterOperator extends BaseFilterOperator {
+  private static final String OPERATOR_NAME = "SortedIndexBasedFilterOperator";
+
+  private final PredicateEvaluator _predicateEvaluator;
+  private final SortedIndexReader _sortedIndexReader;
+  private final int _numDocs;
+
+  SortedIndexBasedFilterOperator(PredicateEvaluator predicateEvaluator, DataSource dataSource, int numDocs) {
+    _predicateEvaluator = predicateEvaluator;
+    _sortedIndexReader = (SortedIndexReader) dataSource.getInvertedIndex();
+    _numDocs = numDocs;
+  }
+
+  @Override
+  protected FilterBlock getNextBlock() {
+    // At this point, we need to create a list of matching docId ranges. There are two kinds of operators:
+    //
+    // - "Additive" operators, such as EQ, IN and RANGE build up a list of ranges and merge overlapping/adjacent ones,
+    //   clipping the ranges to [startDocId; endDocId]
+    //
+    // - "Subtractive" operators, such as NEQ and NOT IN build up a list of ranges that do not match and build a list of
+    //   matching intervals by subtracting a list of non-matching intervals from the given range of
+    //   [startDocId; endDocId]
+    //
+    // For now, we don't look at the cardinality of the column's dictionary, although we should do that if someone
```

Comment:
(nit): should just be cardinality of the column (since size of dictionary is equal to cardinality)

Reply:
I didn't change this block of comments, but seems it identifies it as a new class. Let me update it.